cfg/config.go

```diff
@@ -0,0 +1,184 @@
+// from the given URL.
+func (c *Config) Load() error {
+	config, err := c.backend.Load(c.url)
+	if err != nil {
+		return err
+	}
+	c.Repositories = config.Repositories
+	return nil
+}
+
+// Backend currently being used.
+func (c *Config) Backend() ConfigBackend {
+	return c.backend
+}
+
+// SetURL updates the configuration URL.
+//
+// Next time the config is loaded or saved
+// the new URL will be used.
+func (c *Config) SetURL(u string) error {
+	url, err := url.Parse(u)
+	if err != nil {
+		return err
+	}
+
+	c.url = url
+	return nil
+}
+
+// URL currently being used.
+func (c *Config) URL() *url.URL {
+	return c.url
+}
+
+// New returns a new Config struct
+//
+// The URL will be matched agains all the supported backends and the first
+// backend that can handle the URL scheme will be loaded.
+func New(url string) (*Config, error) {
+	config := &Config{}
+	var backend ConfigBackend
+
+	if url == "" {
+		return nil, fmt.Errorf("Empty URL provided but not supported")
+	}
+
+	err := config.SetURL(url)
+	if err != nil {
+		return nil, err
+	}
+
+	switch config.url.Scheme {
+	case "", "file":
+		if ok, _ := IsEncrypted(config.url); ok {
+			fmt.Println("Loading encrypted configuration file")
+			backend, err = NewAESBackend(config.url)
+			if err != nil {
+				log.Fatalf("error loading the AES configuration backend. err: %v", err)
+			}
+		} else {
+			backend = NewFileBackend()
+		}
+	case "mem":
+		backend = NewMemBackend()
+	case "crypto":
+		if backend, err = NewAESBackend(config.url); err != nil {
+			fmt.Printf("error loading the AES configuration backend. err: %v", err)
+		}
+	default:
+		return nil, fmt.Errorf("Configuration backend '%s' not supported", config.url.Scheme)
+	}
+
+	config.backend = backend
+	return config, nil
+}
+
+// DefaultPath returns Knoxite's default config path.
+//
+// The path returned is OS dependant. If there's an error
+// while trying to figure out the OS dependant path, "knoxite.conf"
+// in the current working dir is returned.
+func DefaultPath() string {
+	userScope := gap.NewScope(gap.User, appName)
+	path, err := userScope.ConfigPath(cfgFileName)
+	if err != nil {
+		return cfgFileName
+	}
+
+	return path
+}
+
+// Lookup tries to find the config file.
+//
+// If a config file is found in the current working directory, that's returned.
+// Otherwise we try to locate it following an OS dependant:
+//
+// Unix:
+//   - ~/.config/app/filename.conf
+// macOS:
+//   - ~/Library/Preferences/app/filename.conf
+// Windows:
+//   - %LOCALAPPDATA%/app/Config/filename.conf
+//
+// If no valid config file is found, an empty string is returned.
+func Lookup() string {
+	paths := []string{}
+	defaultPath := DefaultPath()
+	if exist(defaultPath) {
+		paths = append(paths, defaultPath)
+	}
+
+	// Prepend ./knoxite.conf to the search path if exists, takes priority
+	// over the rest.
+	cwd, err := os.Getwd()
+	if err != nil {
+		log.Printf("Error getting current working directory. err: %v", err)
+		cwd = "."
```

Comment:
Is there really a case where os.Getwd() won't work but opening a file from this directory would work?

Reply:
I'm not sure if there exists a situation where opening a file would work when `os.Getwd()` didn't. But throwing the error here will leave us with a more precise error message as the program failed while calling `os.Getwd()` and not while opening a file.